mountpoint-s3/tests/reftests/harness.rs

```diff
@@ -47,11 +63,81 @@ impl DirectoryIndex {
     }
 }
 
+/// An index into the children of a directory
+#[derive(Debug, Clone, Copy, Arbitrary)]
+pub struct ChildIndex(usize);
+
+impl ChildIndex {
+    /// Get the name and node of the child at the given index in the reference directory (wrapping
+    /// around if the index is larger than the number of children). Returns None if the reference
+    /// directory is empty.
+    fn get<'a>(&self, reference: &'a BTreeMap<String, Node>) -> Option<(&'a str, &'a Node)> {
+        if reference.is_empty() {
+            None
+        } else {
+            let idx = self.0 % reference.len();
+            let key = reference.keys().nth(idx).unwrap();
+            Some((key, reference.get(key).unwrap()))
+        }
+    }
+}
+
+/// An index into the reference model's list of inflight writes. We use this to randomly select an
+/// inflight write to operate on.
+#[derive(Debug, Clone, Copy, Arbitrary)]
+pub struct InflightWriteIndex(usize);
+
+#[derive(Debug)]
+pub struct InflightWrite {
+    path: PathBuf,
+    inode: InodeNo,
+    /// Initially None until the file is opened by [Op::StartWriting]
+    file_handle: Option<u64>,
+    object: MockObject,
+    written: usize,
+}
+
+#[derive(Debug, Default)]
+struct InflightWrites {
+    writes: Vec<InflightWrite>,
+}
+
+impl InflightWrites {
+    fn index(&self, index: InflightWriteIndex) -> Option<usize> {
+        let inflight_writes_len = self.writes.len();
+        if inflight_writes_len == 0 {
+            None
+        } else {
+            Some(index.0 % inflight_writes_len)
+        }
+    }
+
+    fn insert(&mut self, write: InflightWrite) -> InflightWriteIndex {
+        self.writes.push(write);
+        InflightWriteIndex(self.writes.len() - 1)
+    }
+
+    fn get(&self, index: InflightWriteIndex) -> Option<&InflightWrite> {
+        Some(&self.writes[self.index(index)?])
+    }
+
+    fn get_mut(&mut self, index: InflightWriteIndex) -> Option<&mut InflightWrite> {
+        let index = self.index(index)?;
+        Some(&mut self.writes[index])
+    }
+
+    fn remove(&mut self, index: InflightWriteIndex) -> InflightWrite {
+        let index = self.index(index).unwrap();
+        self.writes.remove(index)
+    }
+}
+
 #[derive(Debug)]
 pub struct Harness {
     readdir_limit: usize, // max number of entries that a readdir will return; 0 means no limit
     reference: Reference,
     fs: S3Filesystem<Arc<MockClient>, ThreadPool>,
+    inflight_writes: InflightWrites,
```

Comment:
I'm also trying to understand where "local" directories will fit in this.

Reply:
My thinking is that `Reference` models "the file system tree", whereas `Harness` models "the work we're doing". So, for example, `InflightWrites` itself shouldn't be in the reference because it contains open file handles, which aren't part of the tree. But a local file _is_ part of the file system tree, which is why `File::Local` shows up in the reference.

Local directories are interesting because they're functionally not much different to remote directories, just a map from names to entries. So they definitely appear in the `Reference` since they are part of the file system tree. But we also want to check special properties about local directories (e.g. that all their contents are local too). It's not clear to me whether that means we need to track any additional state or not; I think we'll work that out when we do it.